# import #

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split

# fnc #

In [16]:
#debug
def train_test_debug():
    if train_set["text"][0] == training_data1["text"][0]:
        print("train set is ***1***")
    
    if train_set["text"][0] == training_data2["text"][0]:
        print("train set is ***2***")
        
    if test_set["text"][0] == testing_data1["text"][0]:
        print("test set is ***1***")
    
    if test_set["text"][0] == testing_data2["text"][0]:
        print("test set is ***2***")

In [3]:
def clean_text(text):
    text = str(text)
    if "," in text:
        return str(text.replace(",",""))
    if "." in text:
        return str(text.replace(".",""))
    if "।" in text:
        return str(text.replace("।",""))
    if "|" in text:
        return str(text.replace("|",""))
    return str(text)

In [4]:
def get_emotion(emotion):
    if emotion == "happy":
        return 0
    elif emotion == "sad":
        return 1
    elif emotion == "angry":
        return 2
    elif emotion == "disgust":
        return 3
    elif emotion == "fear":
        return 4

# data #

**dataset1**

In [5]:
training_data1 = pd.read_excel("all_data/training_DS1.xlsx")
print(training_data1.groupby("reaction").apply(list))
print(f"shape = {training_data1.shape}")
print(f"dtype = {training_data1.dtypes}")
print(f"value count = \n{training_data1['reaction'].value_counts()}")

training_list1 = training_data1.values.tolist()
print("\n")
testing_data1 = pd.read_excel("all_data/testing_DS1.xlsx")
print(testing_data1.groupby("reaction").apply(list))
print(f"shape = {testing_data1.shape}")
print(f"dtype = {testing_data1.dtypes}")
print(f"value count = \n{testing_data1['reaction'].value_counts()}")

testing_list1 = training_data1.values.tolist()

reaction
angry      [text, reaction]
disgust    [text, reaction]
fear       [text, reaction]
happy      [text, reaction]
sad        [text, reaction]
dtype: object
shape = (3096, 2)
dtype = text        object
reaction    object
dtype: object
value count = 
happy      1086
sad         721
angry       720
disgust     357
fear        212
Name: reaction, dtype: int64


reaction
angry      [text, reaction]
disgust    [text, reaction]
fear       [text, reaction]
happy      [text, reaction]
sad        [text, reaction]
dtype: object
shape = (1032, 2)
dtype = text        object
reaction    object
dtype: object
value count = 
happy      351
angry      269
sad        224
disgust    112
fear        76
Name: reaction, dtype: int64


In [6]:
training_data1["text"] = training_data1["text"].apply(clean_text)
training_data1["label"] = training_data1["reaction"].apply(get_emotion)
training_data1["num_of_words"] = training_data1["text"].apply(lambda x:len(str(x).split()))

print("\n split \n")

testing_data1["text"] = testing_data1["text"].apply(clean_text)
testing_data1["label"] = testing_data1["reaction"].apply(get_emotion)
testing_data1["num_of_words"] = testing_data1["text"].apply(lambda x:len(str(x).split()))


 split 



**dataset2**

In [7]:
training_data2 = pd.read_excel("all_data/training_DS2.xlsx")
print(training_data2.groupby("reaction").apply(list))
print(f"shape = {training_data2.shape}")
print(f"dtype = {training_data2.dtypes}")
print(f"value count = \n{training_data2['reaction'].value_counts()}")

training_list2 = training_data2.values.tolist()
print("\n")
testing_data2 = pd.read_excel("all_data/testing_DS2.xlsx")
print(testing_data2.groupby("reaction").apply(list))
print(f"shape = {testing_data2.shape}")
print(f"dtype = {testing_data2.dtypes}")
print(f"value count = \n{testing_data2['reaction'].value_counts()}")

testing_list2 = training_data2.values.tolist()

reaction
angry      [text, reaction]
disgust    [text, reaction]
fear       [text, reaction]
happy      [text, reaction]
sad        [text, reaction]
dtype: object
shape = (3720, 2)
dtype = text        object
reaction    object
dtype: object
value count = 
happy      883
sad        810
disgust    799
angry      629
fear       599
Name: reaction, dtype: int64


reaction
angry      [text, reaction]
disgust    [text, reaction]
fear       [text, reaction]
happy      [text, reaction]
sad        [text, reaction]
dtype: object
shape = (1265, 2)
dtype = text        object
reaction    object
dtype: object
value count = 
happy      311
disgust    289
sad        275
angry      208
fear       182
Name: reaction, dtype: int64


In [8]:
training_data2["text"] = training_data2["text"].apply(clean_text)
training_data2["label"] = training_data2["reaction"].apply(get_emotion)
training_data2["num_of_words"] = training_data2["text"].apply(lambda x:len(str(x).split()))
print(training_data2.groupby("reaction").apply(list))

print("\n split \n")

testing_data2["text"] = testing_data2["text"].apply(clean_text)
testing_data2["label"] = testing_data2["reaction"].apply(get_emotion)
testing_data2["num_of_words"] = testing_data2["text"].apply(lambda x:len(str(x).split()))
print(training_data2.groupby("reaction").apply(list))

reaction
angry      [text, reaction, label, num_of_words]
disgust    [text, reaction, label, num_of_words]
fear       [text, reaction, label, num_of_words]
happy      [text, reaction, label, num_of_words]
sad        [text, reaction, label, num_of_words]
dtype: object

 split 

reaction
angry      [text, reaction, label, num_of_words]
disgust    [text, reaction, label, num_of_words]
fear       [text, reaction, label, num_of_words]
happy      [text, reaction, label, num_of_words]
sad        [text, reaction, label, num_of_words]
dtype: object


**select dataset**

In [9]:
train_set = None
test_set = None

In [10]:
def select_train_test(train,test):
    global train_set
    global test_set
    if train == 1 and test == 1:
        print("train = 1 test = 1 loaded")
        print(testing_data1.groupby("reaction").apply(list))
        train_set = training_data1.drop(columns=["reaction","num_of_words"])
        test_set = testing_data1.drop(columns=["reaction","num_of_words"])
    elif train == 2 and test == 2:
        print("train = 2 test = 2 loaded")
        print(training_data2.groupby("reaction").apply(list))
        train_set = training_data2.drop(columns=["reaction","num_of_words"])
        test_set = testing_data2.drop(columns=["reaction","num_of_words"])
    elif train == 1 and test == 2:
        print("train = 1 test = 2 loaded")
        train_set = training_data1.drop(columns=["reaction","num_of_words"])
        test_set = testing_data2.drop(columns=["reaction","num_of_words"])
    elif train == 2 and test == 1:
        print("train = 2 test = 1 loaded")
        train_set = training_data2.drop(columns=["reaction","num_of_words"])
        test_set = testing_data1.drop(columns=["reaction","num_of_words"])
    else:
        print("choose between train 1/2 and test 1/2")
        
    print(f"value count = \n{train_set['label'].value_counts()}")
    print(f"value count = \n{test_set['label'].value_counts()}")


In [11]:
from simpletransformers.classification import ClassificationModel

# Select train test #

In [12]:
select_train_test(1,1)
train_test_debug()

train = 1 test = 1 loaded
reaction
angry      [text, reaction, label, num_of_words]
disgust    [text, reaction, label, num_of_words]
fear       [text, reaction, label, num_of_words]
happy      [text, reaction, label, num_of_words]
sad        [text, reaction, label, num_of_words]
dtype: object
value count = 
0    1086
1     721
2     720
3     357
4     212
Name: label, dtype: int64
value count = 
0    351
2    269
1    224
3    112
4     76
Name: label, dtype: int64
train set is ***1***
test set is ***1***


In [16]:
select_train_test(2,2)
train_test_debug()

train = 2 test = 2 loaded
reaction
angry      [text, reaction, label, num_of_words]
disgust    [text, reaction, label, num_of_words]
fear       [text, reaction, label, num_of_words]
happy      [text, reaction, label, num_of_words]
sad        [text, reaction, label, num_of_words]
dtype: object
value count = 
0    883
1    810
3    799
2    629
4    599
Name: label, dtype: int64
value count = 
0    311
3    289
1    275
2    208
4    182
Name: label, dtype: int64
train set is ***2***
test set is ***2***


In [12]:
select_train_test(1,2)
train_test_debug()

train = 1 test = 2 loaded
value count = 
0    1086
1     721
2     720
3     357
4     212
Name: label, dtype: int64
value count = 
0    311
3    289
1    275
2    208
4    182
Name: label, dtype: int64
train set is ***1***
test set is ***2***


In [17]:
select_train_test(2,1)
train_test_debug()

train = 2 test = 1 loaded
value count = 
0    883
1    810
3    799
2    629
4    599
Name: label, dtype: int64
value count = 
0    351
2    269
1    224
3    112
4     76
Name: label, dtype: int64
train set is ***2***
test set is ***1***


# Model "monsoon-nlp/bangla-electra" #

In [21]:
model = ClassificationModel('bert', 'monsoon-nlp/bangla-electra', num_labels=5, use_cuda=True, args={
    'reprocess_input_data': True,
    'use_cached_eval_features': False,
    'overwrite_output_dir': True,
    'num_train_epochs': 3,
    'silent': True
}) # , weight=[2.5, 1.0]
model.train_model(train_set.sample(frac=1))

You are using a model of type electra to instantiate a model of type bert. This is not supported for all configurations of models and can yield errors.
Some weights of the model checkpoint at monsoon-nlp/bangla-electra were not used when initializing BertForSequenceClassification: ['electra.embeddings.LayerNorm.bias', 'electra.encoder.layer.11.attention.self.query.bias', 'electra.encoder.layer.4.attention.output.dense.bias', 'electra.encoder.layer.2.attention.output.dense.bias', 'electra.encoder.layer.6.output.LayerNorm.weight', 'electra.encoder.layer.1.attention.self.value.weight', 'electra.encoder.layer.3.attention.self.key.weight', 'electra.encoder.layer.2.attention.self.query.bias', 'electra.encoder.layer.4.output.LayerNorm.weight', 'electra.encoder.layer.8.attention.self.value.bias', 'electra.encoder.layer.10.attention.self.query.bias', 'electra.embeddings_project.weight', 'electra.encoder.layer.4.output.LayerNorm.bias', 'electra.encoder.layer.11.attention.self.value.bias', 'elect

Some weights of BertForSequenceClassification were not initialized from the model checkpoint at monsoon-nlp/bangla-electra and are newly initialized: ['encoder.layer.9.output.dense.bias', 'encoder.layer.4.output.dense.bias', 'encoder.layer.4.attention.self.value.bias', 'encoder.layer.5.attention.output.LayerNorm.bias', 'encoder.layer.0.attention.self.query.weight', 'encoder.layer.8.output.LayerNorm.weight', 'encoder.layer.11.attention.self.key.bias', 'encoder.layer.1.intermediate.dense.weight', 'encoder.layer.5.attention.self.key.weight', 'encoder.layer.8.output.LayerNorm.bias', 'encoder.layer.8.attention.output.LayerNorm.weight', 'encoder.layer.6.attention.output.dense.bias', 'embeddings.LayerNorm.bias', 'encoder.layer.1.attention.self.value.weight', 'encoder.layer.7.attention.self.key.bias', 'encoder.layer.1.attention.self.key.weight', 'encoder.layer.3.attention.output.dense.weight', 'encoder.layer.1.attention.output.dense.weight', 'encoder.layer.6.attention.output.dense.weight', 'en

The tokenizer class you load from this checkpoint is not the same type as the class this function is called from. It may result in unexpected tokenization. 
The tokenizer class you load from this checkpoint is 'ElectraTokenizer'. 
The class this function is called from is 'BertTokenizer'.
The tokenizer class you load from this checkpoint is not the same type as the class this function is called from. It may result in unexpected tokenization. 
The tokenizer class you load from this checkpoint is 'ElectraTokenizer'. 
The class this function is called from is 'BertTokenizerFast'.
C:\Users\meera\anaconda3\lib\site-packages\simpletransformers\classification\classification_model.py:612: UserWarning: Dataframe headers not specified. Falling back to using column 0 as text and column 1 as labels.
  warnings.warn(


(1161, 1.2837424791649843)

In [22]:
result, model_outputs, wrong_predictions = model.eval_model(test_set)
bads = {}
for pred in wrong_predictions:
    if pred.label in bads:
        bads[pred.label] += 1
    else:
        bads[pred.label] = 1
print("wrong predictions:")
print(str(len(wrong_predictions)) + ' wrong out of ' + str(len(test_set)))
print(f"accuracy = {(len(test_set)-len(wrong_predictions))/len(test_set)*100}")
bads

C:\Users\meera\anaconda3\lib\site-packages\simpletransformers\classification\classification_model.py:1454: UserWarning: Dataframe headers not specified. Falling back to using column 0 as text and column 1 as labels.
  warnings.warn(


wrong predictions:
837 wrong out of 1265
accuracy = 33.83399209486166


{3: 289, 4: 182, 1: 114, 2: 139, 0: 113}

In [23]:
train_test_debug()

train set is ***1***
test set is ***2***


# Report monsoon-nlp/bangla-electra #

train set is ***1***
test set is ***1***
wrong predictions:
474 wrong out of 1032
accuracy = 54.06976744186046

train set is ***2***
test set is ***2***
wrong predictions:
628 wrong out of 1265
accuracy = 50.35573122529644

train set is ***1***
test set is ***2***
wrong predictions:
874 wrong out of 1265
accuracy = 30.909090909090907

train set is ***2***
test set is ***1***
wrong predictions:
699 wrong out of 1032
accuracy = 32.26744186046512

# Model "sagorsarker/bangla-bert-base" #

In [18]:
train_test_debug()

train set is ***2***
test set is ***1***


In [19]:
bert2 = ClassificationModel('bert', 'sagorsarker/bangla-bert-base', num_labels=5, use_cuda=True, args={
    'reprocess_input_data': True,
    'use_cached_eval_features': False,
    'overwrite_output_dir': True,
    'num_train_epochs': 3,
    'silent': True
})
bert2.train_model(train_set.sample(frac=1))

Some weights of the model checkpoint at sagorsarker/bangla-bert-base were not used when initializing BertForSequenceClassification: ['cls.predictions.decoder.weight', 'cls.predictions.transform.LayerNorm.bias', 'cls.seq_relationship.bias', 'cls.predictions.transform.dense.weight', 'cls.predictions.transform.dense.bias', 'cls.predictions.bias', 'cls.predictions.transform.LayerNorm.weight', 'cls.seq_relationship.weight']
- This IS expected if you are initializing BertForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of BertForSequenceClassification were not initialized from the model ch

(1395, 0.7065620677018251)

In [20]:
result, model_outputs, wrong_predictions = bert2.eval_model(test_set)
bads = {}
for pred in wrong_predictions:
    if pred.label in bads:
        bads[pred.label] += 1
    else:
        bads[pred.label] = 1
print("wrong predictions:")
print(str(len(wrong_predictions)) + ' wrong out of ' + str(len(test_set)))
print(f"accuracy = {(len(test_set)-len(wrong_predictions))/len(test_set)*100}")
bads

C:\Users\meera\anaconda3\lib\site-packages\simpletransformers\classification\classification_model.py:1454: UserWarning: Dataframe headers not specified. Falling back to using column 0 as text and column 1 as labels.
  warnings.warn(


wrong predictions:
578 wrong out of 1032
accuracy = 43.992248062015506


{0: 132, 1: 184, 2: 143, 4: 73, 3: 46}

In [21]:
train_test_debug()

train set is ***2***
test set is ***1***


# Report sagorsarker/bangla-bert-base #

train set is ***1***
test set is ***1***
wrong predictions:
214 wrong out of 1032
accuracy = 79.26356589147287

train set is ***2***
test set is ***2***
wrong predictions:
185 wrong out of 1265
accuracy = 85.37549407114624

train set is ***1***
test set is ***2***
wrong predictions:
722 wrong out of 1265
accuracy = 42.92490118577075

train set is ***2***
test set is ***1***
wrong predictions:
578 wrong out of 1032
accuracy = 43.992248062015506